File: boilerplate/java/order-service/src/main/java/com/example/orderservice/infrastructure/logging/SecurityAuditLogger.java
package com.example.orderservice.infrastructure.logging;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

/**
 * Security audit logger for standardized security event logging.
 * 
 * Logs all security-relevant events in structured JSON format
 * with MDC context (userId, ipAddress, sessionId, userAgent).
 * 
 * Events logged:
 * - AUTH_SUCCESS / AUTH_FAILURE
 * - AUTHZ_FAILURE
 * - SENSITIVE_DATA_ACCESS
 * - SECURITY_CONFIG_CHANGE
 * - SESSION_REVOKED
 * - PASSWORD_RESET_REQUESTED
 * - MFA_ENABLED / MFA_DISABLED
 */
@Component
public class SecurityAuditLogger {
    
    private static final Logger auditLogger = LoggerFactory.getLogger("SECURITY_AUDIT");
    
    public enum SecurityEventType {
        AUTH_SUCCESS,
        AUTH_FAILURE,
        AUTHZ_FAILURE,
        SENSITIVE_DATA_ACCESS,
        SECURITY_CONFIG_CHANGE,
        SESSION_REVOKED,
        PASSWORD_RESET_REQUESTED,
        MFA_ENABLED,
        MFA_DISABLED,
        MFA_VERIFICATION_SUCCESS,
        MFA_VERIFICATION_FAILURE,
        RATE_LIMIT_EXCEEDED,
        ACCOUNT_LOCKED,
        ACCOUNT_UNLOCKED
    }
    
    /**
     * Log authentication success.
     */
    public void logAuthenticationSuccess(String userId, String ipAddress, String userAgent) {
        logSecurityEvent(SecurityEventType.AUTH_SUCCESS, Map.of(
            "userId", userId,
            "ipAddress", ipAddress,
            "userAgent", userAgent,
            "message", "User authenticated successfully"
        ));
    }
    
    /**
     * Log authentication failure.
     */
    public void logAuthenticationFailure(String username, String ipAddress, String reason) {
        logSecurityEvent(SecurityEventType.AUTH_FAILURE, Map.of(
            "username", username,
            "ipAddress", ipAddress,
            "reason", reason,
            "message", "Authentication failed"
        ));
    }
    
    /**
     * Log authorization failure (access denied).
     */
    public void logAuthorizationFailure(String userId, String resource, String action) {
        logSecurityEvent(SecurityEventType.AUTHZ_FAILURE, Map.of(
            "userId", userId,
            "resource", resource,
            "action", action,
            "message", "Unauthorized access attempt"
        ));
    }
    
    /**
     * Log access to sensitive data (PII, PH).
     */
    public void logSensitiveDataAccess(String userId, String dataType, String recordId) {
        logSecurityEvent(SecurityEventType.SENSITIVE_DATA_ACCESS, Map.of(
            "userId", userId,
            "dataType", dataType,
            "recordId", recordId,
            "message", "Sensitive data accessed"
        ));
    }
    
    /**
     * Log security configuration changes.
     */
    public void logSecurityConfigChange(String userId, String configType, String oldValue, String newValue) {
        logSecurityEvent(SecurityEventType.SECURITY_CONFIG_CHANGE, Map.of(
            "userId", userId,
            "configType", configType,
            "oldValue", maskSensitiveData(oldValue),
            "newValue", maskSensitiveData(newValue),
            "message", "Security configuration changed"
        ));
    }
    
    /**
     * Log session revocation (security reason).
     */
    public void logSessionRevoked(String userId, String sessionId, String reason) {
        logSecurityEvent(SecurityEventType.SESSION_REVOKED, Map.of(
            "userId", userId,
            "sessionId", sessionId,
            "reason", reason,
            "message", "Session revoked for security reason"
        ));
    }
    
    /**
     * Log password reset request.
     */
    public void logPasswordResetRequested(String username, String ipAddress) {
        logSecurityEvent(SecurityEventType.PASSWORD_RESET_REQUESTED, Map.of(
            "username", username,
            "ipAddress", ipAddress,
            "message", "Password reset requested"
        ));
    }
    
    /**
     * Log MFA enabled.
     */
    public void logMfaEnabled(String userId, String method) {
        logSecurityEvent(SecurityEventType.MFA_ENABLED, Map.of(
            "userId", userId,
            "method", method,
            "message", "MFA enabled"
        ));
    }
    
    /**
     * Log MFA disabled.
     */
    public void logMfaDisabled(String userId, String method) {
        logSecurityEvent(SecurityEventType.MFA_DISABLED, Map.of(
            "userId", userId,
            "method", method,
            "message", "MFA disabled"
        ));
    }
    
    /**
     * Log rate limit exceeded.
     */
    public void logRateLimitExceeded(String ipAddress, String endpoint, long retryAfterSeconds) {
        logSecurityEvent(SecurityEventType.RATE_LIMIT_EXCEEDED, Map.of(
            "ipAddress", ipAddress,
            "endpoint", endpoint,
            "retryAfterSeconds", retryAfterSeconds,
            "message", "Rate limit exceeded"
        ));
    }
    
    /**
     * Log account lockout.
     */
    public void logAccountLocked(String username, String ipAddress, int failedAttempts) {
        logSecurityEvent(SecurityEventType.ACCOUNT_LOCKED, Map.of(
            "username", username,
            "ipAddress", ipAddress,
            "failedAttempts", failedAttempts,
            "message", "Account locked due to failed login attempts"
        ));
    }
    
    /**
     * Generic security event logging with MDC context.
     */
    private void logSecurityEvent(SecurityEventType eventType, Map<String, Object> eventData) {
        try {
            // Add MDC context
            MDC.put("event_type", eventType.name());
            MDC.put("timestamp", Instant.now().toString());
            
            // Add user context if available
            String userId = MDC.get("userId");
            if (userId != null) {
                MDC.put("userId", userId);
            }
            
            // Build structured log message
            Map<String, Object> logEntry = new HashMap<>(eventData);
            logEntry.put("eventType", eventType.name());
            logEntry.put("timestamp", Instant.now().toString());
            
            // Log based on severity
            switch (eventType) {
                case AUTH_FAILURE:
                case AUTHZ_FAILURE:
                case ACCOUNT_LOCKED:
                case RATE_LIMIT_EXCEEDED:
                    auditLogger.warn("SECURITY_EVENT: {}", toJson(logEntry));
                    break;
                case SECURITY_CONFIG_CHANGE:
                case SESSION_REVOKED:
                    auditLogger.error("SECURITY_EVENT: {}", toJson(logEntry));
                    break;
                default:
                    auditLogger.info("SECURITY_EVENT: {}", toJson(logEntry));
            }
            
        } finally {
            // Clear MDC
            MDC.clear();
        }
    }
    
    /**
     * Convert map to JSON string.
     */
    private String toJson(Map<String, Object> data) {
        StringBuilder json = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : data.entrySet()) {
            if (!first) json.append(",");
            json.append("\"").append(entry.getKey()).append("\":");
            if (entry.getValue() instanceof String) {
                json.append("\"").append(entry.getValue()).append("\"");
            } else {
                json.append(entry.getValue());
            }
            first = false;
        }
        json.append("}");
        return json.toString();
    }
    
    /**
     * Mask sensitive data in logs.
     */
    private String maskSensitiveData(String data) {
        if (data == null || data.length() < 4) {
            return "***";
        }
        return data.substring(0, 2) + "***" + data.substring(data.length() - 2);
    }
}

File: tools/cookiecutter-python/{{cookiecutter.project_slug}}/src/infrastructure/logging/security_audit_logger.py
"""
Security audit logger for standardized security event logging.

Logs all security-relevant events in structured JSON format
with context (user_id, ip_address, session_id, user_agent).

Events logged:
- AUTH_SUCCESS / AUTH_FAILURE
- AUTHZ_FAILURE
- SENSITIVE_DATA_ACCESS
- SECURITY_CONFIG_CHANGE
- SESSION_REVOKED
- PASSWORD_RESET_REQUESTED
- MFA_ENABLED / MFA_DISABLED
- RATE_LIMIT_EXCEEDED
- ACCOUNT_LOCKED
"""

import json
import logging
from datetime import datetime, timezone
from enum import Enum
from typing import Optional, Dict, Any


class SecurityEventType(str, Enum):
    """Security event types for audit logging."""
    AUTH_SUCCESS = "AUTH_SUCCESS"
    AUTH_FAILURE = "AUTH_FAILURE"
    AUTHZ_FAILURE = "AUTHZ_FAILURE"
    SENSITIVE_DATA_ACCESS = "SENSITIVE_DATA_ACCESS"
    SECURITY_CONFIG_CHANGE = "SECURITY_CONFIG_CHANGE"
    SESSION_REVOKED = "SESSION_REVOKED"
    PASSWORD_RESET_REQUESTED = "PASSWORD_RESET_REQUESTED"
    MFA_ENABLED = "MFA_ENABLED"
    MFA_DISABLED = "MFA_DISABLED"
    MFA_VERIFICATION_SUCCESS = "MFA_VERIFICATION_SUCCESS"
    MFA_VERIFICATION_FAILURE = "MFA_VERIFICATION_FAILURE"
    RATE_LIMIT_EXCEEDED = "RATE_LIMIT_EXCEEDED"
    ACCOUNT_LOCKED = "ACCOUNT_LOCKED"
    ACCOUNT_UNLOCKED = "ACCOUNT_UNLOCKED"


# Configure dedicated security audit logger
audit_logger = logging.getLogger("SECURITY_AUDIT")
audit_logger.setLevel(logging.INFO)

# Add handler if not already present
if not audit_logger.handlers:
    handler = logging.StreamHandler()
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )
    handler.setFormatter(formatter)
    audit_logger.addHandler(handler)


class SecurityAuditLogger:
    """
    Security audit logger for standardized security event logging.
    
    Usage:
        logger = SecurityAuditLogger()
        logger.log_authentication_success("user-123", "192.168.1.1")
        logger.log_authorization_failure("user-123", "/orders/456", "GET")
    """
    
    def __init__(self, context: Optional[Dict[str, str]] = None):
        """
        Initialize security audit logger.
        
        Args:
            context: Optional context dict with user_id, ip_address, session_id, etc.
        """
        self.context = context or {}
    
    def _log_security_event(
        self,
        event_type: SecurityEventType,
        event_data: Dict[str, Any],
        level: int = logging.INFO
    ) -> None:
        """
        Log security event with structured data.
        
        Args:
            event_type: Type of security event
            event_data: Event data dictionary
            level: Logging level (INFO, WARNING, ERROR)
        """
        # Build log entry
        log_entry = {
            "eventType": event_type.value,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            **self.context,
            **event_data
        }
        
        # Mask sensitive data
        log_entry = self._mask_sensitive_fields(log_entry)
        
        # Log event
        audit_logger.log(level, f"SECURITY_EVENT: {json.dumps(log_entry)}")
    
    def log_authentication_success(
        self,
        user_id: str,
        ip_address: str,
        user_agent: Optional[str] = None
    ) -> None:
        """Log authentication success."""
        self._log_security_event(
            SecurityEventType.AUTH_SUCCESS,
            {
                "userId": user_id,
                "ipAddress": ip_address,
                "userAgent": user_agent,
                "message": "User authenticated successfully"
            }
        )
    
    def log_authentication_failure(
        self,
        username: str,
        ip_address: str,
        reason: str
    ) -> None:
        """Log authentication failure."""
        self._log_security_event(
            SecurityEventType.AUTH_FAILURE,
            {
                "username": username,
                "ipAddress": ip_address,
                "reason": reason,
                "message": "Authentication failed"
            },
            level=logging.WARNING
        )
    
    def log_authorization_failure(
        self,
        user_id: str,
        resource: str,
        action: str
    ) -> None:
        """Log authorization failure (access denied)."""
        self._log_security_event(
            SecurityEventType.AUTHZ_FAILURE,
            {
                "userId": user_id,
                "resource": resource,
                "action": action,
                "message": "Unauthorized access attempt"
            },
            level=logging.WARNING
        )
    
    def log_sensitive_data_access(
        self,
        user_id: str,
        data_type: str,
        record_id: str
    ) -> None:
        """Log access to sensitive data (PII, PHI)."""
        self._log_security_event(
            SecurityEventType.SENSITIVE_DATA_ACCESS,
            {
                "userId": user_id,
                "dataType": data_type,
                "recordId": record_id,
                "message": "Sensitive data accessed"
            }
        )
    
    def log_security_config_change(
        self,
        user_id: str,
        config_type: str,
        old_value: Any,
        new_value: Any
    ) -> None:
        """Log security configuration changes."""
        self._log_security_event(
            SecurityEventType.SECURITY_CONFIG_CHANGE,
            {
                "userId": user_id,
                "configType": config_type,
                "oldValue": self._mask_sensitive_data(str(old_value)),
                "newValue": self._mask_sensitive_data(str(new_value)),
                "message": "Security configuration changed"
            },
            level=logging.ERROR
        )
    
    def log_session_revoked(
        self,
        user_id: str,
        session_id: str,
        reason: str
    ) -> None:
        """Log session revocation (security reason)."""
        self._log_security_event(
            SecurityEventType.SESSION_REVOKED,
            {
                "userId": user_id,
                "sessionId": session_id,
                "reason": reason,
                "message": "Session revoked for security reason"
            },
            level=logging.WARNING
        )
    
    def log_password_reset_requested(
        self,
        username: str,
        ip_address: str
    ) -> None:
        """Log password reset request."""
        self._log_security_event(
            SecurityEventType.PASSWORD_RESET_REQUESTED,
            {
                "username": username,
                "ipAddress": ip_address,
                "message": "Password reset requested"
            }
        )
    
    def log_mfa_enabled(
        self,
        user_id: str,
        method: str
    ) -> None:
        """Log MFA enabled."""
        self._log_security_event(
            SecurityEventType.MFA_ENABLED,
            {
                "userId": user_id,
                "method": method,
                "message": "MFA enabled"
            }
        )
    
    def log_mfa_disabled(
        self,
        user_id: str,
        method: str
    ) -> None:
        """Log MFA disabled."""
        self._log_security_event(
            SecurityEventType.MFA_DISABLED,
            {
                "userId": user_id,
                "method": method,
                "message": "MFA disabled"
            }
        )
    
    def log_rate_limit_exceeded(
        self,
        ip_address: str,
        endpoint: str,
        retry_after_seconds: int
    ) -> None:
        """Log rate limit exceeded."""
        self._log_security_event(
            SecurityEventType.RATE_LIMIT_EXCEEDED,
            {
                "ipAddress": ip_address,
                "endpoint": endpoint,
                "retryAfterSeconds": retry_after_seconds,
                "message": "Rate limit exceeded"
            },
            level=logging.WARNING
        )
    
    def log_account_locked(
        self,
        username: str,
        ip_address: str,
        failed_attempts: int
    ) -> None:
        """Log account lockout."""
        self._log_security_event(
            SecurityEventType.ACCOUNT_LOCKED,
            {
                "username": username,
                "ipAddress": ip_address,
                "failedAttempts": failed_attempts,
                "message": "Account locked due to failed login attempts"
            },
            level=logging.ERROR
        )
    
    @staticmethod
    def _mask_sensitive_data(data: str) -> str:
        """Mask sensitive data in logs."""
        if not data or len(data) < 4:
            return "***"
        return f"{data[:2]}***{data[-2:]}"
    
    def _mask_sensitive_fields(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Mask sensitive fields in log entry."""
        sensitive_fields = ["secret", "password", "token", "authorization", "credential"]
        masked = data.copy()
        
        for field in sensitive_fields:
            for key in masked.keys():
                if field.lower() in key.lower():
                    masked[key] = self._mask_sensitive_data(str(masked[key]))
        
        return masked
